Постановка целей

licosty opened the pull request:
Реализованы основные необходимые операции:
- создать новую цель
- обновить цель
- удалить цель
- получить все цели по фильтру

licosty committed:
goals: create, update, delete, filters

licosty committed:
goals: add exception handler

itechnik-swd commented on src/main/java/school/faang/user_service/controller/goal/GoalController.java:
```diff
@@ -0,0 +1,63 @@
+        validateString(createGoalDto.title(), "title");
+        validateString(createGoalDto.description(), "description");
+        validateNotNull(createGoalDto.userIds(), "userIds");
+        return ResponseEntity.ok(goalService.create(createGoalDto));
```
•	ok() соответствует коду 200;
•	created() — коду 201;
Т.к. метод `create()`, то можно `ok()` заменить на `created()`.

licosty replied:
Поскольку остальные методы не возвращают значение ResponseEntity, привела этот метод к общему виду. Поэтому поменяла статус через аннотацию)

licosty committed:
goals: controller - create
- изменено возвращаемое значение;
- http-status изменен на 201-created

avzag1 commented on src/main/java/school/faang/user_service/controller/goal/GoalController.java:
```diff
@@ -0,0 +1,65 @@
+        return goalService.getByFilters(filters);
+    }
+
+    private void validateString(String value, String paramName) {
```
Для валидации входящих данных на контроллер не лучше ли использовать аннотации?

licosty replied:
@avzag1 Лучше, конечно. Когда по спринтам мы дойдем до этой темы, думаю можно будет доработать)

avzag1 commented on src/main/java/school/faang/user_service/controller/goal/GoalController.java:
```diff
@@ -0,0 +1,65 @@
+        }
+    }
+
+    private void validateNotNull(Object value, String paramName) {
```
То же самое. Для валидации входящих данных на контроллер не лучше ли использовать аннотации?

licosty replied:
@avzag1 Лучше, конечно. Когда по спринтам мы дойдем до этой темы, думаю можно будет доработать)

avzag1 commented on src/main/java/school/faang/user_service/dto/goal/CreateGoalDto.java:
```diff
@@ -0,0 +1,12 @@
+import java.time.LocalDateTime;
+import java.util.List;
+
+public record CreateGoalDto(Long parentId,
```
Не лучше ли использовать одну общую DTO, чем 2 разных?

licosty replied:
@avzag1 Это было условием задачи. Но в целом да, в данном случае одной можно было обойтись

avzag1 commented on src/main/java/school/faang/user_service/repository/goal/GoalRepository.java:
```diff
@@ -32,6 +35,26 @@ SELECT COUNT(ug.goal_id) FROM user_goal ug
+
+    @Transactional
     @Modifying
     @Query(nativeQuery = true, value = """
```
Зачем использовать нативный SQL, если можно использовать кастомные методы jpa?

licosty replied:
@avzag1 Какой кастомный метод можно было использовать в данном случае?

avzag1 commented on src/main/java/school/faang/user_service/repository/goal/GoalRepository.java:
```diff
@@ -64,9 +87,22 @@ WITH RECURSIVE subtasks AS (
+            JOIN FETCH g.users
+            WHERE g.id = :goalId
+            """)
+    Optional<Goal> getGoalWithUsersById(long goalId);
```
Зачем используешь нативный SQL вместо jpa?

licosty replied:
@avzag1 Чтобы более тонко настроить получаемый результат. 
Если в данном случае ты знаешь как заменить на чистый Jpa, прошу подсказать

avzag1 commented on src/main/java/school/faang/user_service/repository/goal/GoalRepository.java:
```diff
@@ -64,9 +87,22 @@ WITH RECURSIVE subtasks AS (
+    Optional<Goal> getGoalWithUsersById(long goalId);
+
     default Goal getByIdOrThrow(long goalId) {
         return findById(goalId).orElseThrow(
```
Не лучше ли это вынести в сервисный слой?

licosty replied:
@avzag1 Так не думаю. Это метод по получению данных из базы, в нем нет бизнес-логики

zaqqaz226 commented on src/main/java/school/faang/user_service/controller/goal/GoalController.java:
```diff
@@ -0,0 +1,65 @@
+        return goalService.getByFilters(filters);
+    }
+
+    private void validateString(String value, String paramName) {
```
Эти 2 проверки вообще можно сдвинуть в одну. Просто добавь вторую проверку в if с логическим или.

licosty replied:
Сделала, но еще с проверкой на строку

zaqqaz226 commented on src/main/java/school/faang/user_service/exception/handler/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,53 @@
+
+@Slf4j
+@RestControllerAdvice
+public class GlobalExceptionHandler {
```
![image](https://github.com/user-attachments/assets/1be5012d-036d-4205-9200-7bbdd511f968)

zaqqaz226 commented on src/main/java/school/faang/user_service/exception/handler/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,53 @@
+    @ExceptionHandler(EntityNotFoundException.class)
+    public ErrorResponse handleEntityNotFound(EntityNotFoundException e) {
+        log.error(e.getMessage(), e);
+        return new ErrorResponse("Entity not found", e.getMessage(), LocalDateTime.now());
```
А зачем тебе timestamp в исключении? Это все будет видно в логах)

licosty replied:
Подумала, что дополнительно тоже неплохо вернуть. Но да, вероятно это лишнее

zaqqaz226 replied:
Точно лишнее)

zaqqaz226 commented on src/main/java/school/faang/user_service/service/goal/GoalServiceImpl.java:
```diff
@@ -0,0 +1,177 @@
+@Slf4j
+@Service
+@RequiredArgsConstructor
+@ConfigurationProperties(prefix = "goal")
```
Эта аннотация используется чуть иначе) Она для того, чтобы сконфигурировать класс под пропертя. https://www.baeldung.com/configuration-properties-in-spring-boot Так что тут либо используй только @Value, либо уже скорфи класс для использования пропертей. Я за второй вариант)

licosty replied:
Если правильно поняла, то вынесла в отдельный проперти класс GoalProperty

zaqqaz226 replied:
Все правильно)

zaqqaz226 commented on src/main/java/school/faang/user_service/service/goal/GoalServiceImpl.java:
```diff
@@ -0,0 +1,177 @@
+    private boolean isParticipant(List<User> users, long currentId) {
+        return users != null
+               && users.stream()
+                       .anyMatch(user -> user.getId().equals(currentId));
```
Примитивы все таки лучше сравнить через "==". А то если там где то обертка, то тут можно проблемы с null получить)

licosty replied:
В общем-то у user как раз обертка. Если обернуть currentId в Long, это поможет избежать ошибки даже если оба null, но это не хорошее решение. Поэтому сделала через Objects.equals()

licosty committed:
goals:
- добавлен GoalProperty
- объединена валидация полей ы GoalController
- убрана дата из ErrorResponse

licosty committed:
goals: fix default value properties

zaqqaz226 commented on src/main/java/school/faang/user_service/config/property/GoalProperty.java:
```diff
@@ -0,0 +1,9 @@
+import org.springframework.boot.context.properties.bind.DefaultValue;
+
+@ConfigurationProperties(prefix = "goal")
+public record GoalProperty(@DefaultValue("1") int minParticipantsCount,
```
Просто отлично!)

zaqqaz226 commented:
Ну что ж, принимай поздравления\соболезнования, можно делать тесты))

licosty committed:
goals: тесты на GoalService
- в update добавлена проверка, - новый ментор не участник

zaqqaz226 commented on src/test/java/school/faang/user_service/service/goal/GoalServiceImplTest.java:
```diff
@@ -0,0 +1,466 @@
+    @Test
+    @DisplayName("Успешное создание цели ментром")
+    void positive_whenMentor_shouldCreateGoal() {
+        CreateGoalDto createGoalDto =
```
Горит сарай, гори и хата! Давай тогда и создание этой сущности в отдельный метод вынесем)

licosty replied:
Сделала)

zaqqaz226 commented on src/test/java/school/faang/user_service/service/goal/GoalServiceImplTest.java:
```diff
@@ -0,0 +1,466 @@
+    @Test
+    @DisplayName("Успешное создание цели самим пользователем")
+    void positive_whenRegularUser_shouldCreateGoal() {
+        CreateGoalDto createGoalDto =
```
А то вот тут он второй раз встречается

licosty replied:
Готово

zaqqaz226 commented on src/test/java/school/faang/user_service/service/goal/GoalServiceImplTest.java:
```diff
@@ -0,0 +1,466 @@
+        assertEquals(expected, actual);
+    }
+
+    @ParameterizedTest
```
<img width="544" height="400" alt="image" src="https://github.com/user-attachments/assets/b62ecd5b-542c-4d13-9c38-fe2fd012343f" />

zaqqaz226 commented on src/test/java/school/faang/user_service/service/goal/GoalServiceImplTest.java:
```diff
@@ -0,0 +1,466 @@
+    }
+
+    @ParameterizedTest
+    @MethodSource("provideUsersHaveNoCreateAuthority")
```
Единственное что, я бы сурсы для параметризации ставил бы прям над тестом. Чтобы сразу было видно, что будем туда подавать. Но это вкусовщина и можно оставить и так как есть.

licosty replied:
@zaqqaz226 Но если правильно поняла, в таких параметрах никак не могу использовать константы

zaqqaz226 commented on src/test/java/school/faang/user_service/service/goal/GoalServiceImplTest.java:
```diff
@@ -0,0 +1,466 @@
+    @MethodSource("provideUsersHaveNoCreateAuthority")
+    @DisplayName("Ошибка создания цели - недостаточно прав")
+    void negative_whenHasNoAuthority_shouldNotCreateAndThrowException(Long mentorId, Long userId, Long currentId) {
+        CreateGoalDto createGoalDto =
```
Опять же я б в отдельный приватный метод создание вынес.

licosty replied:
Сделано

licosty committed:
goals: tests refactoring

licosty committed:
goals: test добавлено сравнение текста ошибки
